src/shallow/ShallowWrapper.lua

```diff
@@ -0,0 +1,159 @@
+local RoactRoot = script.Parent.Parent
+
+local Children = require(RoactRoot.PropMarkers.Children)
+local ElementKind = require(RoactRoot.ElementKind)
+local ElementUtils = require(RoactRoot.ElementUtils)
+local strict = require(RoactRoot.strict)
+local Symbol = require(RoactRoot.Symbol)
+local Snapshot = require(script.Parent.Snapshot)
+local VirtualNodeConstraints = require(script.Parent.VirtualNodeConstraints)
+
+local InternalData = Symbol.named("InternalData")
+
+local ShallowWrapper = {}
+local ShallowWrapperPublic = {}
+local ShallowWrapperMetatable = {
+	__index = ShallowWrapperPublic,
+}
+
+local function getComponentFromVirtualNode(virtualNode)
+	local element = virtualNode.currentElement
+	local kind = ElementKind.of(element)
+
+	if kind == ElementKind.Host
+		or kind == ElementKind.Function
+		or kind == ElementKind.Stateful
+	then
+		return element.component
+	else
+		error(("shallow wrapper does not support element of kind %q"):format(tostring(kind)))
+	end
+end
+
+local function findNextVirtualNode(virtualNode, maxDepth)
+	local currentDepth = 0
+	local currentNode = virtualNode
+	local nextNode = currentNode.children[ElementUtils.UseParentKey]
+
+	while currentDepth < maxDepth and nextNode ~= nil do
+		currentNode = nextNode
+		nextNode = currentNode.children[ElementUtils.UseParentKey]
+		currentDepth = currentDepth + 1
+	end
+
+	return currentNode
+end
+
+local function getChildren(virtualNode, results, maxDepth)
```

Comment:
I wanted to avoid copying table here because of the recursion. The way it's working right now, each child will add itself to the list, so it means that if make tables immutable I would need to create at least one list per child and recombine them. So are you sure it would be worth it to make it immutable here?